### PR TITLE
Box tools for AI Agents

### PR DESCRIPTION
**Pull request by barduinor**
# Description

Suite of Box tools for seamlessly integrating Box content with Llama Index. These tools offer an agnostic interface, allowing them to be used within any agent built within the Llama Index framework.

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No (New package)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/tools/llama-index-tools-box/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+check-filenames = true
+check-hidden = true
+# Feel free to un-skip examples, and experimental, you will just need to
+# work through many typos (--write-changes and --interactive will help)
+skip = "*.csv,*.html,*.json,*.jsonl,*.pdf,*.txt,*.ipynb"
+
+[tool.llamahub]
+contains_example = true
+import_path = "llama_index.tools.box"
+
```
Comment (nerdai) [inline llama-index-integrations/tools/llama-index-tools-box/pyproject.toml:15]: Please add `tool.llamahub.class_authors` section:

For example:
```
[tool.llamahub.class_authors]
BoxAIExtractToolSpec = "llama-index"
BoxAIPromptToolSpec = "llama-index"
BoxSearchToolSpec = "llama-index"
BoxSearchByMetadataToolSpec = "llama-index"
BoxTextExtractToolSpec = "llama-index"
```

Feel free to replace `llama-index` above with your Github username.

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    def __init__(
+        self, box_client: BoxClient, options: BoxSearchByMetadataOptions
+    ) -> None:
+        self._box_client = box_client
+        self._options = options
+
+    def search(
+        self,
+        # query: Optional[str] = None,
+        # query_params: Optional[Dict[str, str]] = None,
```
Comment (nerdai) [inline llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py:56]: can we please remove commented code?

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_prompt/base.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            ai_prompt=ai_prompt,
+        )
+
+        for payload in payloads:
+            file = payload.resource_info
+            ai_response = payload.ai_response
+
+            # create a document
+            doc = Document(
+                # id=file.id,
```
Comment (nerdai) [inline llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_prompt/base.py:48]: can we please remove commented code?

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_extract/base.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            ai_prompt=ai_prompt,
+        )
+
+        for payload in payloads:
+            file = payload.resource_info
+            ai_response = payload.ai_response
+
+            # create a document
+            doc = Document(
+                # id=file.id,
```
Comment (nerdai) [inline llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_extract/base.py:48]: can we please remove commented code?

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_prompt/base.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+class BoxAIPromptToolSpec(BaseToolSpec):
+    spec_functions = ["ai_prompt"]
+
+    _box_client: BoxClient
+
+    def __init__(self, box_client: BoxClient) -> None:
+        self._box_client = box_client
+
+    def ai_prompt(
```
Comment (nerdai) [inline llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_prompt/base.py:23]: Can we add a nice doc string for this?

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_extract/base.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+class BoxAIExtractToolSpec(BaseToolSpec):
+    spec_functions = ["ai_extract"]
+
+    _box_client: BoxClient
+
+    def __init__(self, box_client: BoxClient) -> None:
+        self._box_client = box_client
+
+    def ai_extract(
```
Comment (nerdai) [inline llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_extract/base.py:23]: Please add a nice doc string for this spec_function

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    get_box_files_payload,
+    search_files_by_metadata,
+)
+
+
+class BoxSearchByMetadataOptions:
+    from_: str
+    ancestor_folder_id: str
+    query: Optional[str] = (None,)
+    # query_params: Optional[Dict[str, str]] = (None,)
```
Comment (nerdai) [inline llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py:21]: can we please remove commented code?

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        # query_params: Optional[Dict[str, str]] = None,
+        query_params: Optional[str] = None,
+    ) -> List[Document]:
+        """
+        Searches for Box resources based on metadata and returns a list of Llama Index
+        Documents.
+
+        This method utilizes the Box API search functionality to find resources
+        matching the provided metadata query. It then returns a list containing the IDs
+        of the found resources.
```
Comment (nerdai) [inline llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py:65]: Can we add a line detailing args?

**Review by nerdai — COMMENTED** [review]
Thanks @barduinor!

I left a few comments, mostly minor in nature. Otherwise the PR looks pretty good.

For doc strings of all `spec-functions`, we should use a nice doc string format. For example, "google" format:

```python
    def my_function(self, param: str, param2: str) -> str:
        """_summary_

        Args:
            param (str): _description_
            param2 (str): _description_

        Returns:
            str: _description_
        """
```

File: llama-index-integrations/tools/llama-index-tools-box/pyproject.toml (hunk 1 above)
Comment (barduinor) [reply llama-index-integrations/tools/llama-index-tools-box/pyproject.toml:15]: Added

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py (hunk 2 above)
Comment (barduinor) [reply llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py:56]: Removed

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_prompt/base.py (hunk 3 above)
Comment (barduinor) [reply llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_prompt/base.py:48]: Removed

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_extract/base.py (hunk 4 above)
Comment (barduinor) [reply llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_extract/base.py:48]: Removed

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py (hunk 7 above)
Comment (barduinor) [reply llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py:21]: Removed

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_prompt/base.py (hunk 5 above)
Comment (barduinor) [reply llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_prompt/base.py:50]: added

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_extract/base.py (hunk 6 above)
Comment (barduinor) [reply llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/ai_extract/base.py:50]: added

File: llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py (hunk 8 above)
Comment (barduinor) [reply llama-index-integrations/tools/llama-index-tools-box/llama_index/tools/box/search_by_metadata/base.py]: added

**Comment by barduinor** [comment]
@nerdai , thank you for your feedback.

I believe I covered all your requests, removed commented code and added doc strings.

Much appreciated.

**Review by nerdai — APPROVED** [review]
Thanks @barduinor for making the changes -- lgtm!

**Pull request merged**